add API for re-functionalization an aten op; other functionalization fixes

bdhirsh opened the pull request:
I moved out the changes to `FunctionalTensorWrapper.h` from the LTC <> functionalization PR into a separate PR here, so dealing with XLA failures will be a bit easier.

Specifically, the LTC PR will make a few operators like `pixel_shuffle` that are functional, but decompose into view ops, require re-functionalization once they hit the XLA backend. This PR exposes a helper utility to do that through `functionalize_aten_op`. This PR also contains the changes to:
- fix `detach()` for `FunctionalTensorWrapper`
- fix some undefined tensor handling cases


I have an XLA patch here to do the re-functionalizing: https://github.com/pytorch/xla/pull/3646

Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #75527
* __->__ #79420

bdhirsh committed:
add API for re-functionalization an aten op; other functionalization fixes

[ghstack-poisoned]

ezyang reviewed (approved):
I've already reviewed this right

bdhirsh committed:
Update on "add API for re-functionalization an aten op; other functionalization fixes"

I moved out the changes to `FunctionalTensorWrapper.h` from the LTC <> functionalization PR into a separate PR here, so dealing with XLA failures will be a bit easier.

Specifically, the LTC PR will make a few operators like `pixel_shuffle` that are functional, but decompose into view ops, require re-functionalization once they hit the XLA backend. This PR exposes a helper utility to do that through `functionalize_aten_op`. This PR also contains the changes to:
- fix `detach()` for `FunctionalTensorWrapper`
- fix some undefined tensor handling cases


I have an XLA patch here to do the re-functionalizing: https://github.com/pytorch/xla/pull/3646




[ghstack-poisoned]

bdhirsh committed:
Update on "add API for re-functionalization an aten op; other functionalization fixes"

I moved out the changes to `FunctionalTensorWrapper.h` from the LTC <> functionalization PR into a separate PR here, so dealing with XLA failures will be a bit easier.

Specifically, the LTC PR will make a few operators like `pixel_shuffle` that are functional, but decompose into view ops, require re-functionalization once they hit the XLA backend. This PR exposes a helper utility to do that through `functionalize_aten_op`. This PR also contains the changes to:
- fix `detach()` for `FunctionalTensorWrapper`
- fix some undefined tensor handling cases


I have an XLA patch here to do the re-functionalizing: https://github.com/pytorch/xla/pull/3646




[ghstack-poisoned]

bdhirsh committed:
Update on "add API for re-functionalization an aten op; other functionalization fixes"

I moved out the changes to `FunctionalTensorWrapper.h` from the LTC <> functionalization PR into a separate PR here, so dealing with XLA failures will be a bit easier.

Specifically, the LTC PR will make a few operators like `pixel_shuffle` that are functional, but decompose into view ops, require re-functionalization once they hit the XLA backend. This PR exposes a helper utility to do that through `functionalize_aten_op`. This PR also contains the changes to:
- fix `detach()` for `FunctionalTensorWrapper`
- fix some undefined tensor handling cases


I have an XLA patch here to do the re-functionalizing: https://github.com/pytorch/xla/pull/3646




[ghstack-poisoned]

bdhirsh committed:
Update on "add API for re-functionalization an aten op; other functionalization fixes"

I moved out the changes to `FunctionalTensorWrapper.h` from the LTC <> functionalization PR into a separate PR here, so dealing with XLA failures will be a bit easier.

Specifically, the LTC PR will make a few operators like `pixel_shuffle` that are functional, but decompose into view ops, require re-functionalization once they hit the XLA backend. This PR exposes a helper utility to do that through `functionalize_aten_op`. This PR also contains the changes to:
- fix `detach()` for `FunctionalTensorWrapper`
- fix some undefined tensor handling cases


I have an XLA patch here to do the re-functionalizing: https://github.com/pytorch/xla/pull/3646




[ghstack-poisoned]

bdhirsh committed:
Update on "add API for re-functionalization an aten op; other functionalization fixes"

I moved out the changes to `FunctionalTensorWrapper.h` from the LTC <> functionalization PR into a separate PR here, so dealing with XLA failures will be a bit easier.

Specifically, the LTC PR will make a few operators like `pixel_shuffle` that are functional, but decompose into view ops, require re-functionalization once they hit the XLA backend. This PR exposes a helper utility to do that through `functionalize_aten_op`. This PR also contains the changes to:
- fix `detach()` for `FunctionalTensorWrapper`
- fix some undefined tensor handling cases


I have an XLA patch here to do the re-functionalizing: https://github.com/pytorch/xla/pull/3646




[ghstack-poisoned]

bdhirsh committed:
Update on "add API for re-functionalization an aten op; other functionalization fixes"

I moved out the changes to `FunctionalTensorWrapper.h` from the LTC <> functionalization PR into a separate PR here, so dealing with XLA failures will be a bit easier.

Specifically, the LTC PR will make a few operators like `pixel_shuffle` that are functional, but decompose into view ops, require re-functionalization once they hit the XLA backend. This PR exposes a helper utility to do that through `functionalize_aten_op`. This PR also contains the changes to:
- fix `detach()` for `FunctionalTensorWrapper`
- fix some undefined tensor handling cases


I have an XLA patch here to do the re-functionalizing: https://github.com/pytorch/xla/pull/3646




[ghstack-poisoned]

bdhirsh committed:
Update on "add API for re-functionalization an aten op; other functionalization fixes"

I moved out the changes to `FunctionalTensorWrapper.h` from the LTC <> functionalization PR into a separate PR here, so dealing with XLA failures will be a bit easier.

Specifically, the LTC PR will make a few operators like `pixel_shuffle` that are functional, but decompose into view ops, require re-functionalization once they hit the XLA backend. This PR exposes a helper utility to do that through `functionalize_aten_op`. This PR also contains the changes to:
- fix `detach()` for `FunctionalTensorWrapper`
- fix some undefined tensor handling cases


I have an XLA patch here to do the re-functionalizing: https://github.com/pytorch/xla/pull/3646




[ghstack-poisoned]

bdhirsh committed:
Update on "add API for re-functionalization an aten op; other functionalization fixes"

I moved out the changes to `FunctionalTensorWrapper.h` from the LTC <> functionalization PR into a separate PR here, so dealing with XLA failures will be a bit easier.

Specifically, the LTC PR will make a few operators like `pixel_shuffle` that are functional, but decompose into view ops, require re-functionalization once they hit the XLA backend. This PR exposes a helper utility to do that through `functionalize_aten_op`. This PR also contains the changes to:
- fix `detach()` for `FunctionalTensorWrapper`
- fix some undefined tensor handling cases


I have an XLA patch here to do the re-functionalizing: https://github.com/pytorch/xla/pull/3646




[ghstack-poisoned]

bdhirsh committed:
Update on "add API for re-functionalization an aten op; other functionalization fixes"

I moved out the changes to `FunctionalTensorWrapper.h` from the LTC <> functionalization PR into a separate PR here, so dealing with XLA failures will be a bit easier.

Specifically, the LTC PR will make a few operators like `pixel_shuffle` that are functional, but decompose into view ops, require re-functionalization once they hit the XLA backend. This PR exposes a helper utility to do that through `functionalize_aten_op`. This PR also contains the changes to:
- fix `detach()` for `FunctionalTensorWrapper`
- fix some undefined tensor handling cases


I have an XLA patch here to do the re-functionalizing: https://github.com/pytorch/xla/pull/3646




[ghstack-poisoned]

pytorchmergebot committed:
Update on "add API for re-functionalization an aten op; other functionalization fixes"

I moved out the changes to `FunctionalTensorWrapper.h` from the LTC <> functionalization PR into a separate PR here, so dealing with XLA failures will be a bit easier.

Specifically, the LTC PR will make a few operators like `pixel_shuffle` that are functional, but decompose into view ops, require re-functionalization once they hit the XLA backend. This PR exposes a helper utility to do that through `functionalize_aten_op`. This PR also contains the changes to:
- fix `detach()` for `FunctionalTensorWrapper`
- fix some undefined tensor handling cases


I have an XLA patch here to do the re-functionalizing: https://github.com/pytorch/xla/pull/3646




[ghstack-poisoned]

bdhirsh commented:
closing, a version of this was already landed
